release/src-rt-5.04behnd.4916/bcmdrivers/broadcom/net/wl/impl103/main/components/opensource/router_tools/libnl/man/nl-qdisc-add.8
.TH nl\-qdisc 8 "21 October 2010" "libnl"
.SH NAME
nl\-qdisc\-{add|list|delete} - Manage queueing disciplines
.SH SYNOPSIS
.B nl\-qdisc\-add \-\-dev
.I dev
.B \-\-parent
.I id
.B [OPTIONS]
.I qdisc-type
.B [QDISC]
.sp
.B nl\-qdisc\-delete [ \-\-interactive ] [OPTIONS]
.sp
.B nl\-qdisc\-list [OPTIONS]

.SH DESCRIPTION
.PP
The nl\-qdisc tools allow to manage and configure queueing disciplines
(qdiscs) in the kernel.

.SH OPTIONS
.TP
.BR \-\^h " or " \-\-help
Print help text to console and exit.
.TP
.BR \-\^v " or " \-\-version
Print versioning information to console and exit.
.TP
.BR \-\^q " or " \-\-quiet
Do not print informal notifications about actions taken to the console.
By default a short description of each qdisc added/update/deleted will
be printed to the console. This option disables this behaviour.
.TP
.BR \-\^d " or " \-\-dev "=DEV"
Network device the qdisc is attached to.
.TP
.BR \-\^p " or " \-\-parent "=ID"
Identifier of the parent qdisc/class this qdisc is attached to. The
identifier can be specified as classid, name or one of the special
values "root" or "ingress".
.TP
.BR \-\^i " or " \-\-id "=ID"
Identifier of qdisc. It can be specified as classid or name.

.SS nl\-qdisc\-add Options
.TP
.B \-\-update
Update qdisc if it already exists, otherwise attempting to add a qdisc which already
exists will result in an error. This does not include changing the type of the qdisc,
use \-\-replace if you wish to do so.
.TP
.B \-\-replace
Replace or update qdisc if it already exists. Same behaviour as \-\-update but will
completely replace the qdisc if it exists already.
.TP
.B \-\-update\-only
Update an existing qdisc but do not create it if it does not exist.
.TP
.B \-\-replace\-only
Update or replace an existing qdisc but do not create it if it does exist.

.SS nl\-qdisc\-delete Options
.TP
.B \-\-interactive
The interactive mode requires confirmation by the user for each qdisc deleted. It
will print a prompt for each qdisc matching the provided filter and requires the
user to answer 'y'es or 'n'o.
.TP
.B \-\-yes
Make the default answer for interactive prompts be 'y'es. This option is also
required to delete all qdiscs on all network devices.
.TP
.BR \-\^k " or " \-\-kind "=TYPE"
Only delete qdiscs of this type.

.SS nl\-qdisc\-list Options
.TP
.B \-\-details
Show detailed information for each qdisc listed.
.TP
.B \-\-stats
Show statistics information for each qdisc listed. This option will also turn
on detailed information automatically.
.TP
.BR \-\^r " or " \-\-recursive
List all TC objects recurisvely attached to all qdiscs matching the filter.
.TP
.BR \-\^k " or " \-\-kind "=TYPE"
Only list qdiscs of this type.

.SH USAGE
.PP
Add a HTB root qdisc with id "5:":
.PP
.RS
nl\-qdisc\-add \-\-dev eth0 \-\-parent root \-\-id 5: htb
.RE
.PP
List all qdiscs on eth0 and print statistical data:
.PP
.RS
nl\-qdisc\-list \-\-stats \-\-dev eth0
.RE
.PP
Delete the qdisc "5:":
.RS
nl\-qdisc\-delete \-\-id 5:
.RE

.SH "SEE ALSO"
.PP
.B nl\-classid\-lookup(8)

.SH AUTHOR
.PP
Thomas Graf is the original author and current maintainer of libnl and
libnl tools. Many people have contributed to it since.


release/src-rt-5.04behnd.4916/bcmdrivers/broadcom/net/wl/impl103/main/components/opensource/router_tools/libnl/src/nl-qdisc-add.c
/*
 * src/nl-qdisc-add.c     Add Queueing Discipline
 *
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation version 2.1
 *	of the License.
 *
 * Copyright (c) 2010 Thomas Graf <tgraf@suug.ch>
 */

#include <netlink/cli/utils.h>
#include <netlink/cli/tc.h>
#include <netlink/cli/qdisc.h>
#include <netlink/cli/link.h>

#include <netlink-private/route/tc-api.h>

static int quiet = 0;

static void print_usage(void)
{
	printf(
"Usage: nl-qdisc-add [OPTIONS]... QDISC [CONFIGURATION]...\n"
"\n"
"OPTIONS\n"
" -q, --quiet               Do not print informal notifications.\n"
" -h, --help                Show this help text.\n"
" -v, --version             Show versioning information.\n"
"     --update              Update qdisc if it exists.\n"
"     --replace             Replace or update qdisc if it exists.\n"
"     --update-only         Only update qdisc, never create it.\n"
"     --replace-only        Only replace or update qdisc, never create it.\n"
" -d, --dev=DEV             Network device the qdisc should be attached to.\n"
" -i, --id=ID               ID of new qdisc (default: auto-generated)r\n"
" -p, --parent=ID           ID of parent { root | ingress | QDISC-ID }\n"
"\n"
"CONFIGURATION\n"
" -h, --help                Show help text of qdisc specific options.\n"
"\n"
"EXAMPLE\n"
"   $ nl-qdisc-add --dev=eth1 --parent=root htb --rate=100mbit\n"
"\n"
	);
	exit(0);
}

int main(int argc, char *argv[])
{
	struct nl_sock *sock;
	struct rtnl_qdisc *qdisc;
	struct rtnl_tc *tc;
	struct nl_cache *link_cache;
	struct nl_dump_params dp = {
		.dp_type = NL_DUMP_DETAILS,
		.dp_fd = stdout,
	};
	struct nl_cli_tc_module *tm;
	struct rtnl_tc_ops *ops;
	int err, flags = NLM_F_CREATE | NLM_F_EXCL;
	char *kind, *id = NULL;

	sock = nl_cli_alloc_socket();
	nl_cli_connect(sock, NETLINK_ROUTE);

	link_cache = nl_cli_link_alloc_cache(sock);

 	qdisc = nl_cli_qdisc_alloc();
	tc = (struct rtnl_tc *) qdisc;

	for (;;) {
		int c, optidx = 0;
		enum {
			ARG_REPLACE = 257,
			ARG_UPDATE = 258,
			ARG_REPLACE_ONLY,
			ARG_UPDATE_ONLY,
		};
		static struct option long_opts[] = {
			{ "quiet", 0, 0, 'q' },
			{ "help", 0, 0, 'h' },
			{ "version", 0, 0, 'v' },
			{ "dev", 1, 0, 'd' },
			{ "parent", 1, 0, 'p' },
			{ "id", 1, 0, 'i' },
			{ "replace", 0, 0, ARG_REPLACE },
			{ "update", 0, 0, ARG_UPDATE },
			{ "replace-only", 0, 0, ARG_REPLACE_ONLY },
			{ "update-only", 0, 0, ARG_UPDATE_ONLY },
			{ 0, 0, 0, 0 }
		};

		c = getopt_long(argc, argv, "+qhvd:p:i:",
				long_opts, &optidx);
		if (c == -1)
			break;

		switch (c) {
		case 'q': quiet = 1; break;
		case 'h': print_usage(); break;
		case 'v': nl_cli_print_version(); break;
		case 'd': nl_cli_tc_parse_dev(tc, link_cache, optarg); break;
		case 'p': nl_cli_tc_parse_parent(tc, optarg); break;
		case 'i': id = strdup(optarg); break;
		case ARG_UPDATE: flags = NLM_F_CREATE; break;
		case ARG_REPLACE: flags = NLM_F_CREATE | NLM_F_REPLACE; break;
		case ARG_UPDATE_ONLY: flags = 0; break;
		case ARG_REPLACE_ONLY: flags = NLM_F_REPLACE; break;
		}
 	}

	if (optind >= argc)
		print_usage();

	if (!rtnl_tc_get_ifindex(tc))
		nl_cli_fatal(EINVAL, "You must specify a network device (--dev=XXX)");

	if (!rtnl_tc_get_parent(tc))
		nl_cli_fatal(EINVAL, "You must specify a parent");

	if (id) {
		nl_cli_tc_parse_handle(tc, id, 1);
		free(id);
	}

	kind = argv[optind++];
	rtnl_tc_set_kind(tc, kind);

	if (!(ops = rtnl_tc_get_ops(tc)))
		nl_cli_fatal(ENOENT, "Unknown qdisc \"%s\"", kind);

	if (!(tm = nl_cli_tc_lookup(ops)))
		nl_cli_fatal(ENOTSUP, "Qdisc type \"%s\" not supported.", kind);

	tm->tm_parse_argv(tc, argc, argv);

	if (!quiet) {
		printf("Adding ");
		nl_object_dump(OBJ_CAST(qdisc), &dp);
 	}

	if ((err = rtnl_qdisc_add(sock, qdisc, flags)) < 0)
		nl_cli_fatal(EINVAL, "Unable to add qdisc: %s", nl_geterror(err));

	return 0;
}
